### ISSUE TITLE
CiliumEndpointSlice : some ces and cep are lost control and cannot be deleted

### ISSUE THREAD

**Issue by orange30** [cilium/cilium#35532]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.14.13 and lower than v1.15.0

### What happened?

Some ces and cep are lost control and cannot be deleted  because of the race condition between "cep informer goroutine" and "worker goroutine"

### How can we reproduce the issue?

1, when a pod is deleted and the pod is the last cep of the ces.  The "worker goroutine" is at:
func (c *CiliumEndpointSliceController) worker
    -> func (c *CiliumEndpointSliceController) processNextWorkItem
        -> func (c *CiliumEndpointSliceController) syncCES
            -> func (r *reconciler) reconcileCESUpdate
                -> func (c *cesMgr) updateCESInCache
                -> func (c *cesMgr) clearRemovedCEPs
Then it is scheduled to other goroutine.

2, we create the same pod again,  the "cep informer goroutine" is at:
func (legacy *legacyOnLeader) onStart
      -> func CiliumEndpointsSliceInit
            -> func CiliumEndpointsInit
                  -> func endpointUpdated
                        -> func (c *cesMgr) InsertCEPInCache  
                            -> after the first time to call addCEPtoCES
Then it is scheduled to other goroutine.

3, scheduled to "worker goroutime" again, and then:
func (c *CiliumEndpointSliceController) worker
    -> func (c *CiliumEndpointSliceController) processNextWorkItem
        -> func (c *CiliumEndpointSliceController) syncCES
            -> func (r *reconciler) reconcileCESDelete
                -> func (c *cesMgr) deleteCESFromCache
                    ->func (c *CESToCEPMapping) deleteCES
Then, the management struct between ces and cesTracer has been deleted.
Then it is scheduled to other goroutine.

4, Then it is scheduled to  "cep informer goroutine"  and then "worker goroutime":
In the reconcileCESCreate->updateCESInCache of the "worker goroutine", it will cannot find the cesTracker(because it is delete by the third step).

### Cilium Version

1,  All the version of v1.14, cilium-operator have the problem of this issue.

2, All the version before v.14.0, cilium-operator may crash for a similar rase condition(in the second step it is scheduled before addCEPtoCES).( These versions are no longer maintained. )

### Kernel Version

Any version

### Kubernetes Version

v.1.30

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by orange30** [comment]
@julianwiedmann 
1, I think，for the maintained version, we should fix the problem no matter which method we will use.
2, For the crash problem before v1.14, may be we should mark it to let the user know the version vulnerability at https://github.com/cilium/cilium/security .
For example, in our production environment using verison 1.13, the cilium-operator crashed for this race condition problem, I upload the log here.
[v1.13.11-cilium-operator-crash.txt](https://github.com/user-attachments/files/17517986/v1.13.11-cilium-operator-crash.txt)

**Comment by tamilmani1989** [comment]
@orange30 Does this issue happen for version 1.15 and greater? Also, it would be good if you can you specify the steps to repro this issue

**Comment by jshr-w** [comment]
Yes, +1 on the repro please :) I tried scaling up/down deployments so that CESs were created/deleted, but wasn't able to reproduce the issue, more details would be great. Also, wondering if you are using FCFS or Identity mode?

**Comment by orange30** [comment]
@tamilmani1989 @jshr-w 

1. We are using FCFS mode, but this issue exists in both FCFS and Identity modes for versions earlier than 1.15 .
 
2. For versions earlier than 1.15, in the "cep informer goroutine," the function endpointUpdated calls c.InsertCEPInCache, which uses and creates a desiredCESs entry: c.desiredCESs.getCESTracker(cesName) and c.InsertCEPInCache -> c.createCES -> CESToCEPMapping.insertCES. However, the deletion of the desiredCESs entry via CESToCEPMapping.deleteCES is invoked in the "worker goroutine" through the sequence: CiliumEndpointSliceController.syncCES -> reconciler.reconcileCESDelete -> cesMgr.deleteCESFromCache -> c.desiredCESs.deleteCES. （Here code is from version 1.13.11）
This discrepancy, with creation and deletion occurring in different goroutines, leads to a race condition. 

3.  For versions 1.15 and later, the functions cesManagerFcfs.UpdateCEPMapping, which calls cesMgr.createCES and subsequently CESToCEPMapping.insertCES, as well as cesManagerFcfs.RemoveCEPMapping, which invokes c.mapping.deleteCEP(cepName) and c.mapping.deleteCES(cesName), are all executed within the "cep informer goroutine" (code from the main branch).
 
4.  The reproduction for the issue is quite demanding, and the required scheduling sequence is rather complex. While I currently don't have a better method to reproduce it, this issue has indeed occurred in the production environment. I have uploaded the relevant logs for reference. https://github.com/cilium/cilium/issues/35532#issuecomment-2437127181
